### PR TITLE
Programatic Visulisation API

### PR DESCRIPTION
**Pull request by istefano82**
# Create programatic API for ludwig visualisations.

- Split loading of visualisation data for CLI  from actual visualisation
- All visualisation functions can be used via Python API
- Create VIsualisation Programatic API Integration Test Suite
- Visualisations from CLI now support ground truth test split argument where - 0:train, 1:validation, 2:test split
- Fix bug where  bincounts for gt and predictions  in compare_classifiers_predictions_distribution were not of the same length for rare predictions.
-  For CLI visualisations predictions and probabilities can be used from npy and csv files. Add CLI progression integration tests
- All integration tests are passing OK.

### CODE REVIEW

File: mkdocs/docs/user_guide.md [hunk 1]
```diff
@@ -2664,7 +2666,7 @@ If the values of `subset` is `predictions`, then only datapoints where the the m
 
 ### compare_classifiers_performance_changing_k
 
-This visualization uses the `top_k`, `ground_truth_metadata`, `field`, `probabilities` and `model_names` parameters.
+This visualization uses the `top_k`, `ground_truth`, `field`, `probabilities` and `model_names` parameters.
```
Comment (w4nderlust) [inline mkdocs/docs/user_guide.md]: doublecheck if this is correct

File: mkdocs/docs/user_guide.md [hunk 2]
```diff
@@ -2699,7 +2701,7 @@ Compare Classifier Predictions
 
 ### compare_classifiers_predictions
 
-This visualization uses the `ground_truth`, `ground_truth_metadata`, `field`, `predictions` and `model_names` parameters.
+This visualization uses the `ground_truth`, `ground_truth`, `field`, `predictions` and `model_names` parameters.
```
Comment (w4nderlust) [inline mkdocs/docs/user_guide.md]: double check if this is correct, ground truth is repeted twice

File: mkdocs/docs/user_guide.md [hunk 3]
```diff
@@ -2708,7 +2710,7 @@ This visualization produces a pie chart comparing the predictions of the two mod
 
 ### compare_classifiers_predictions_distribution
 
-This visualization uses the `ground_truth`, `ground_truth_metadata`, `field`, `predictions` and `model_names` parameters.
+This visualization uses the `ground_truth`, `ground_truth`, `field`, `predictions` and `model_names` parameters.
```
Comment (w4nderlust) [inline mkdocs/docs/user_guide.md]: double check if this is correct, ground truth is repeted twice

File: tests/integration_tests/test_api.py [hunk 4]
```diff
@@ -25,7 +25,6 @@
 
 
 # The following imports are pytest fixtures, required for running the tests
-from tests.fixtures.filenames import csv_filename
```
Comment (w4nderlust) [inline tests/integration_tests/test_api.py]: please readd this, pycharm code optimization removed it because it's not referenced, but is actually needed

File: tests/integration_tests/test_experiment.py [hunk 5]
```diff
@@ -41,9 +41,8 @@
 
 from tests.integration_tests.utils import date_feature
 
+
 # The following imports are pytest fixtures, required for running the tests
-from tests.fixtures.filenames import csv_filename
-from tests.fixtures.filenames import yaml_filename
```
Comment (w4nderlust) [inline tests/integration_tests/test_experiment.py]: please readd both lines, pycharm code optimization removed it because it's not referenced, but is actually needed

File: ludwig/utils/visualization_utils.py [hunk 6]
```diff
@@ -39,14 +39,16 @@
...
 logger = logging.getLogger(__name__)
 
 
+
 # plt.rc('xtick', labelsize='x-large')
 # plt.rc('ytick', labelsize='x-large')
 # plt.rc('axes', labelsize='x-large')
 
-def learning_curves_plot(
+def learning_curves_ploload(
```
Comment (w4nderlust) [inline ludwig/utils/visualization_utils.py]: plolead?

File: ludwig/utils/visualization_utils.py [hunk 7]
```diff
@@ -39,14 +39,16 @@
 import ludwig.contrib
 
 
+
```
Comment (w4nderlust) [inline ludwig/utils/visualization_utils.py]: for figuring out the correct spaces please run pycharm's code >reformat code on all sources (not on the test ones, as that messes up with the fictures)

File: ludwig/visualize.py [hunk 8]
```diff
@@ -37,100 +38,729 @@
...
+    :param item: object to be checked and converted
+    :return: original item if it is a list instance or list containing the item.
+    """
+    return item if item is None or isinstance(item, list) else [item]
 
-    training_statistics_per_model_name = [load_json(learning_stats_f)
-                                          for learning_stats_f in
-                                          training_statistics]
 
+def validate_visualisation_prediction_field_from_train_stats(
```
Comment (w4nderlust) [inline ludwig/visualize.py]: please replace every occurrence of "visualisation" with "visualization", adopting the american english convention ;)

File: ludwig/visualize.py [hunk 9]
```diff
@@ -731,34 +1420,45 @@ def compare_classifiers_predictions(
 
 
 def compare_classifiers_predictions_distribution(
-        predictions,
-        ground_truth,
-        field,
+        preds_per_model,
+        gt,
```
Comment (w4nderlust) [inline ludwig/visualize.py]: replcase in all function signatures `preds_per_model`, `probs_per_model` and `gt` with `predictions_per _model`, `probabilities_per_model` and `ground_truth`. As those arguments have docstrings and are exposed, I prefer them to be as explicit as possible. Internally in the functions they can be shortened for brevity, but for the exposed ones I prefer them to be longer and clear.

**Comment by w4nderlust** [comment]
Please @ydudin3 and @msaisumanth take a look at this when you can. I added small comments already, otherwise it looks got to me, but as this is a big PR having at least another pairs of eyes on it would be great.

File: ludwig/api.py [hunk 10]
```diff
@@ -65,6 +65,10 @@
 # logger.setLevel(logging.INFO)
 
 
+logger = logging.getLogger(__name__)
```
Comment (ydudin3) [inline ludwig/api.py]: What is this change relevant to, is this a bug fix?

File: ludwig/utils/data_utils.py [hunk 11]
```diff
@@ -294,22 +294,33 @@ def text_feature_data_field(text_feature):
...
+    """Load experiment data from supported file formats.
+
+    Experiment data can be test/train statistics, model predictions,
+    probability, ground truth,  ground truth metadata.
+    :param file_name: Path to file to be loaded
+    :param field: Target Prediction field.
+    :param dtype:
+    :param ground_truth_split: Ground truth split filter where 0 is train 1 is
+    validation and 2 is test split. By default test split is used when loading
+    ground truth form hdf5.
```
Comment (ydudin3) [inline ludwig/utils/data_utils.py]: `from`

File: tests/integration_tests/test_visualisations.py [hunk 12]
```diff
@@ -250,10 +249,80 @@ def test_visualisation_compare_performance_output_saved(csv_filename):
...
+        )
+    ]
+    output_features = [categorical_feature(vocab_size=2, reduce_input='sum')]
 
-def test_visualisation_compare_classifiers_from_prob_output_saved(csv_filename):
+    # Generate test data
+    rel_path = generate_data(input_features, output_features, csv_filename)
+    encoder = 'cnnrnn'
+    input_features[0]['encoder'] = encoder
+    exp_dir_name = run_experiment(input_features, output_features,
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: Split into one argument per line

File: tests/integration_tests/test_visualisations.py [hunk 13]
```diff
@@ -250,10 +249,80 @@ def test_visualisation_compare_performance_output_saved(csv_filename):
...
+    shutil.rmtree('results', ignore_errors=True)
+    for file in glob.glob(experiment_source_data_name + '.*'):
+        try:
+            os.remove(file)
+        except OSError as e:  # if failed, report it back to the user
+            print("Error: %s - %s." % (e.filename, e.strerror))
+
+
+def test_visualisation_compare_classifiers_from_prob_npy_output_saved(
+        csv_filename
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: wrong indentation?

File: tests/integration_tests/test_visualisations.py [hunk 14]
```diff
@@ -319,10 +384,81 @@ def test_visualisation_compare_classifiers_from_prob_output_saved(csv_filename):
         except OSError as e:  # if failed, report it back to the user
             print("Error: %s - %s." % (e.filename, e.strerror))
 
+def test_visualisation_compare_classifiers_from_pred_npy_output_saved(
+        csv_filename
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: wrong indentation?

File: ludwig/api.py (hunk 10 above)
Comment (w4nderlust) [reply ludwig/api.py]: i guess it's a byproduct of merging from master, maybe merging again from master to be sure about these logger things would be a good idea

File: tests/integration_tests/test_visualisations.py [hunk 15]
```diff
@@ -319,10 +384,81 @@ def test_visualisation_compare_classifiers_from_prob_output_saved(csv_filename):
...
+        categorical_feature(
+            vocab_size=10,
+            loss='sampled_softmax_cross_entropy'
+        )
+    ]
+    output_features = [categorical_feature(vocab_size=2, reduce_input='sum')]
+
+    # Generate test data
+    rel_path = generate_data(input_features, output_features, csv_filename)
+    encoder = 'cnnrnn'
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: This is not used anywhere besides the next line so no need to declare a variable. Just do this instead
`input_features[0]['encoder'] = 'cnnrnn'`

File: tests/integration_tests/test_visualisations.py [hunk 16]
```diff
@@ -319,10 +384,81 @@ def test_visualisation_compare_classifiers_from_prob_output_saved(csv_filename):
...
+    input_features[0]['encoder'] = encoder
+    exp_dir_name = run_experiment(input_features, output_features,
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: One argument per line

File: tests/integration_tests/test_visualisations.py [hunk 17]
```diff
@@ -668,6 +782,72 @@ def test_visualisation_compare_classifiers_predictions_output_saved(
...
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: One arg per line

File: tests/integration_tests/test_visualisations.py [hunk 18]
```diff
@@ -668,6 +782,72 @@ def test_visualisation_compare_classifiers_predictions_output_saved(
...
+        assert 0 == result.returncode
+        assert 1 == len(figure_cnt)
+
+    shutil.rmtree(exp_dir_name, ignore_errors=True)
+    shutil.rmtree('results', ignore_errors=True)
+    for file in glob.glob(experiment_source_data_name + '.*'):
+        try:
+            os.remove(file)
+        except OSError as e:  # if failed, report it back to the user
+            print("Error: %s - %s." % (e.filename, e.strerror))
```
Comment (ydudin3) [inline tests/integration_tests/test_visualisations.py]: Maybe add a little more info here, Error deleting file or something like that?

File: ludwig/visualize.py [hunk 19]
```diff
@@ -37,100 +38,729 @@
...
+    :param threshhold_fields: List of threshhold fields per model
+    :raise: RuntimeError
+    """
+    validation_mapping = {
+        'probabilities': probabilities,
+        'treshhold_fields': treshhold_fields
+    }
+    for item, value in validation_mapping.items():
+        item_len = len(value)
+        if not item_len == 2:
```
Comment (ydudin3) [inline ludwig/visualize.py]: Nit, use `if item_len != 2`

File: mkdocs/docs/user_guide.md (hunk 1 above)
Comment (istefano82) [reply mkdocs/docs/user_guide.md]: I have checked it is correct - The new cli function prototype does not need the ground truth metadata. You can also check the old compare_classifiers_performance_changing function declaration to see that it is not using that argument.
def compare_classifiers_performance_changing_k_cli(
        probabilities,
        ground_truth,
        ground_truth_split,
        field,
        **kwargs
):

File: mkdocs/docs/user_guide.md (hunk 2 above)
Comment (istefano82) [reply mkdocs/docs/user_guide.md]: I have checked it is correct - The new cli function prototype does not need the ground truth metadata. You can also check the old compare_classifiers_predictions function declaration to see that it is not using that argument.
def compare_classifiers_predictions_cli(
        predictions,
        ground_truth,
        ground_truth_split,
        field,
        **kwargs
):

Comment (istefano82) [reply mkdocs/docs/user_guide.md]: I have removed the repeated word.

File: mkdocs/docs/user_guide.md (hunk 3 above)
Comment (istefano82) [reply mkdocs/docs/user_guide.md]: This was the old declration of the function - no metadata used:

def compare_classifiers_predictions_distribution(
        predictions,
        ground_truth,
        field,
        labels_limit,
        model_names=None,
        output_directory=None,
        file_format='pdf',
        **kwargs
):

I have removed the repeated word.

File: tests/integration_tests/test_api.py (hunk 4 above)
Comment (istefano82) [reply tests/integration_tests/test_api.py]: Done.

File: tests/integration_tests/test_experiment.py (hunk 5 above)
Comment (istefano82) [reply tests/integration_tests/test_experiment.py]: Added the lines again.

File: ludwig/utils/visualization_utils.py (hunk 6 above)
Comment (istefano82) [reply ludwig/utils/visualization_utils.py]: Some typo during merger - Good catch! :)

File: ludwig/utils/visualization_utils.py (hunk 7 above)
Comment (istefano82) [reply ludwig/utils/visualization_utils.py]: Done - used Reformat and Auto Ident lines on the entire file.

File: ludwig/api.py (hunk 10 above)
Comment (istefano82) [reply ludwig/api.py]: Reverted the state to the one at Uber  master branch.

```suggestion
logger = logging.getLogger(__name__)


# logger.setLevel(logging.INFO)
```

File: ludwig/utils/data_utils.py (hunk 11 above)
Comment (istefano82) [reply ludwig/utils/data_utils.py]: Fixed typo.

File: tests/integration_tests/test_visualisations.py (hunk 12 above)
Comment (istefano82) [reply tests/integration_tests/test_visualisations.py]: Fixed

File: tests/integration_tests/test_visualisations.py (hunk 16 above)
Comment (istefano82) [reply tests/integration_tests/test_visualisations.py]: Fixed.

File: ludwig/visualize.py (hunk 8 above)
Comment (istefano82) [reply ludwig/visualize.py]: Done.

File: tests/integration_tests/test_visualisations.py (hunk 18 above)
Comment (istefano82) [reply tests/integration_tests/test_visualisations.py]: The e.strerror carries the original OSError exception message. I think thtat should be sufficient information on what went wrong.

File: ludwig/visualize.py (hunk 19 above)
Comment (istefano82) [reply ludwig/visualize.py]: Done
